nautobot/users/api/views.py

```diff
@@ -40,6 +44,53 @@ class GroupViewSet(ModelViewSet):
     filterset_class = filters.GroupFilterSet
 
 
+#
+# REST API tokens
+#
+
+
+class TokenViewSet(ModelViewSet):
+    queryset = RestrictedQuerySet(model=Token).prefetch_related("user")
+    serializer_class = serializers.TokenSerializer
+    filterset_class = filters.TokenFilterSet
+
+    def get_queryset(self):
+        """
+        Limit the non-superusers to their own Tokens.
+        """
+        queryset = super().get_queryset()
+        # Workaround for schema generation (drf_yasg)
+        if getattr(self, "swagger_fake_view", False):
+            return queryset.none()
+        if self.request.user.is_superuser:
+            return queryset
+        return queryset.filter(user=self.request.user)
+
+
+class TokenProvisionView(APIView):
+    """
+    Non-authenticated REST API endpoint via which a user may create a Token.
+    """
+
+    permission_classes = []
+
+    def post(self, request):
+        serializer = serializers.TokenProvisionSerializer(data=request.data)
+        serializer.is_valid()
+
+        # Authenticate the user account based on the provided credentials
+        user = authenticate(request=request, username=serializer.data["username"], password=serializer.data["password"])
```

Comment:
We need to rely on authentication upstream vs. hard-coding it in the `post` method here. This view should already be authenticated by the time this method is called.

Reply:
@jathanism I'm not following, this is explicitly for people who are **not** authenticated.

Reply:
I think the concern was that `authenticate()` might not work for non-local authentication methods (such as remote or SSO). However I *think* this should be a non-issue since `authenticate()` tries all configured AUTHENTICATION_BACKENDS. @jathanism can you confirm?

Reply:
@glennmatthews That clears things up, I checked the [Django docs](https://docs.djangoproject.com/en/3.2/topics/auth/default/#django.contrib.auth.authenticate) and they say

> Use [authenticate()](https://docs.djangoproject.com/en/3.2/topics/auth/default/#django.contrib.auth.authenticate) to verify a set of credentials. It takes credentials as keyword arguments, username and password for the default case, checks them against each [authentication backend](https://docs.djangoproject.com/en/3.2/topics/auth/customizing/#authentication-backends), and returns a [User](https://docs.djangoproject.com/en/3.2/ref/contrib/auth/#django.contrib.auth.models.User) object if the credentials are valid for a backend. If the credentials aren’t valid for any backend or if a backend raises [PermissionDenied](https://docs.djangoproject.com/en/3.2/ref/exceptions/#django.core.exceptions.PermissionDenied), it returns None.

So we should be good.

Reply:
After much consideration I think it is out of scope to handle setups that need SAML auth to be able to support issuing tokens over API.

Reply:
Sorry I am late to the party in responding. It's an anti-pattern to hard-code authentication inside of an action method like `post`. In this case the username/password are explicitly being passed in from the payload which binds this to basic (username/password) authentication only.

From the same documentation:

> This is a low level way to authenticate a set of credentials; for example, it’s used by the [RemoteUserMiddleware](https://docs.djangoproject.com/en/3.2/ref/middleware/#django.contrib.auth.middleware.RemoteUserMiddleware). Unless you are writing your own authentication system, you probably won’t use this.

Additionally, DRF has its own authentication classes designed to work with DRF API views. This authentication check should instead be done using an authentication class and then defined on this view using `authentication_classes = [...]`. 

Please see:  https://www.django-rest-framework.org/api-guide/authentication/#setting-the-authentication-scheme

This is what needs to happen:

### nautobot/users/api/authentication.py

```python
from rest_framework import authentication

from . import serializers


class TokenProvisionAuthentication(authentication.BaseAuthentication):
    def authenticate(self, request):
        serializer = serializers.TokenProvisionSerializer(data=request.data)
        serializer.is_valid(raise_exception=True)

        # Authenticate the user account based on the provided credentials
        user = authenticate(
            request=request,
            username=serializer.validated_data["username"],
            password=serializer.validated_data["password"],
        )
        if user is None:
            raise AuthenticationFailed("Invalid username/password")

        return (user, None)
```

### nautobot/users/api/views.py

```python
from . import authentication


class TokenProvisionView(APIView):
    """
    Non-authenticated REST API endpoint via which a user may create a Token.
    """

    authentication_classes = [authentication.TokenProvisionAuthentication]
    permission_classes = []

    def post(self, request):
        # Create a new Token for the User
        token = Token.objects.create(user=request.user)
        data = serializers.TokenSerializer(token, context={"request": request}).data

        return Response(data, status=HTTP_201_CREATED)
```

You see I moved the serialization and authenticate stuff out of the view entirely into the authentication class. That way the two are not tightly coupled.

Reply:
@jathanism thanks for the detailed reply. I'm not going to lie and say I 100% follow but I see the advantage of separating that auth bit out.

I've basically copied and pasted your code in.

Reply:
Thank you! It's really just about decoupling and making sure we don't tie this implementation to a specific hard-coded username/password authentication case even while that is currently the only one, you can see how we made the view handling not care about authentication at all and just expect it to be handled upstream.